state/application.go

```diff
@@ -920,6 +941,80 @@ func (a *Application) newCharmStorageOps(
 	return checkStorageOps, upgradeStorageOps, []txn.Op{storageConstraintsOp}, nil
 }
 
+func (a *Application) newCharmDeviceOps(
+	ch *Charm,
+	units []*Unit,
+	updatedDeviceConstraints map[string]devices.Constraints,
+) ([]txn.Op, []txn.Op, []txn.Op, error) {
+
+	fail := func(err error) ([]txn.Op, []txn.Op, []txn.Op, error) {
+		return nil, nil, nil, errors.Trace(err)
+	}
+
+	// Check device to ensure no referenced device is removed, or changed
+	// in an incompatible way. We do this before computing the new device
+	// constraints, as incompatible charm changes will otherwise yield
+	// confusing error messages that would suggest the user has supplied
+	// invalid constraints.
+	sb, err := NewDeviceBackend(a.st)
+	if err != nil {
+		return fail(err)
+	}
+	// oldCharm, _, err := a.Charm()
+	// if err != nil {
+	// 	return fail(err)
+	// }
+	// oldMeta := oldCharm.Meta()
+	// checkDeviceOps, err := a.checkDeviceUpgrade(ch.Meta(), oldMeta, units)
+	// if err != nil {
+	// 	return fail(err)
+	// }
+
+	// Create or replace device constraints. We take the existing device
+	// constraints, remove any keys that are no longer referenced by the
+	// charm, and update the constraints that the user has specified.
+	var deviceConstraintsOp txn.Op
+	oldDeviceConstraints, err := a.DeviceConstraints()
+	if err != nil {
+		return fail(err)
+	}
+	newDeviceConstraints := oldDeviceConstraints
+	for name, cons := range updatedDeviceConstraints {
+		newDeviceConstraints[name] = cons
+	}
+	for name := range newDeviceConstraints {
+		if _, ok := ch.Meta().Devices[name]; !ok {
+			logger.Warningf("Ignored invalid device constraint %q - not in charm metadata.", name)
+			delete(newDeviceConstraints, name)
+		}
+	}
+	// if err := addDefaultDeviceConstraints(sb, newDeviceConstraints, ch.Meta()); err != nil {
+	// 	return fail(errors.Annotate(err, "adding default device constraints"))
+	// }
+	if err := validateDeviceConstraints(sb, newDeviceConstraints, ch.Meta()); err != nil {
+		return fail(errors.Annotate(err, "validating device constraints"))
+	}
+	newDeviceConstraintsKey := applicationDeviceConstraintsKey(a.doc.Name, ch.URL())
+	if _, err := readDeviceConstraints(sb.mb, newDeviceConstraintsKey); errors.IsNotFound(err) {
```

Comment:
Didn't we just read the device constraints from a.DeviceConstraints()?

Reply:
ah, yes, u r right